src/Microsoft.FeatureManagement/Targeting/ContextualTargetingFeatureVariantAssigner.cs

```diff
@@ -0,0 +1,128 @@
+
+        /// <summary>
+        /// Creates a targeting contextual feature filter.
+        /// </summary>
+        /// <param name="options">Options controlling the behavior of the targeting evaluation performed by the filter.</param>
+        public ContextualTargetingFeatureVariantAssigner(IOptions<TargetingEvaluationOptions> options)
+        {
+            _options = options?.Value ?? throw new ArgumentNullException(nameof(options));
+        }
+
+        /// <summary>
+        /// Assigns one of the variants configured for a feature based off the provided targeting context.
+        /// </summary>
+        /// <param name="variantAssignmentContext">Contextual information available for use during the assignment process.</param>
+        /// <param name="targetingContext">The targeting context used to determine which variant should be assigned.</param>
+        /// <param name="cancellationToken">The cancellation token to cancel the operation.</param>
+        /// <returns></returns>
+        public ValueTask<FeatureVariant> AssignVariantAsync(FeatureVariantAssignmentContext variantAssignmentContext, ITargetingContext targetingContext, CancellationToken cancellationToken)
+        {
+            FeatureDefinition featureDefinition = variantAssignmentContext.FeatureDefinition;
+
+            if (featureDefinition == null)
+            {
+                return new ValueTask<FeatureVariant>((FeatureVariant)null);
+            }
+
+            FeatureVariant variant = null;
+
+            FeatureVariant defaultVariant = null;
+
+            double cumulativePercentage = 0;
+
+            var cumulativeGroups = new Dictionary<string, double>();
+
+            if (featureDefinition.Variants != null)
+            {
+                foreach (FeatureVariant v in featureDefinition.Variants)
+                {
+                    if (defaultVariant == null && v.Default)
+                    {
+                        defaultVariant = v;
+                    }
+
+                    TargetingFilterSettings targetingSettings = v.AssignmentParameters.Get<TargetingFilterSettings>();
+
+                    if (targetingSettings == null)
+                    {
+                        if (v.Default)
+                        {
+                            //
+                            // Valid to omit audience for default variant
+                            continue;
+                        }
+                        else
+                        {
+                            targetingSettings = new TargetingFilterSettings();
+                        }
+                    }
+
+                    if (!TargetingEvaluator.TryValidateSettings(targetingSettings, out string paramName, out string reason))
+                    {
+                        throw new ArgumentException(reason, paramName);
+                    }
+
+                    AccumulateAudience(targetingSettings.Audience, ref cumulativePercentage, ref cumulativeGroups);
```

Comment:
What's this used for?

Reply:
Hard to tell in GitHub comments. Is this for `AccumulateAudience`?

If so,
It's used in the algorithm to select which group should be assigned to if there are multiple groups with configured percentages.

E.g. take the below setup

```
variant 1
  group 1
    60% users

variant 2
  group 1
    15% users

variant 3
  group 1
    25% users
```

Given a user Bob in group 1, we choose the targeted variant by checking
1. For variant 1, is Bob in 60% of users
2. For variant 2, is Bob in 75% (60 + 15) of users
3. For variant 3, is Bob in 100% (60 + 15 + 25) of users

The helper here calculates the cumulative value in that process.

After taking a look at this piece, there is a small bug to address. I'm also going to add a test since that bug wasn't covered by tests.

Reply:
Yes, it's about `AccumulateAudience`. I'm using the codeflow, so the comment placement can be weird. I didn't see the ref'ed variables used anywhere. That's probably the bugs you mentioned.

This is an interesting problem. Not sure how other systems do this. So the order of variants actually matters.

Reply:
> I didn't see the ref'ed variables used anywhere. That's probably the bugs you mentioned.

`cumulativeGroups` is being used already correctly, but yes `cumulativePercentage` is not being used as it should.

Yes, the order of the variants matter.